File: database/src/model/deck_card.rs
use crate::model::card::Db_Card;
use crate::model::deck::Db_Deck;
use crate::sql_types::deck_card_amount::Db_DeckCardAmount;
use crate::sql_types::num::{Db_CardLocalId, Db_DeckId};
use bon::Builder;
use diesel::prelude::*;
use serde::{Deserialize, Serialize};
use specta::Type;

#[derive(
  Builder,
  Associations,
  Identifiable,
  Insertable,
  Selectable,
  Queryable,
  Clone,
  Debug,
  Serialize,
  Deserialize,
  Type,
)]
#[diesel(table_name = crate::schema::deck_card)]
#[diesel(belongs_to(Db_Deck, foreign_key = deck_id))]
#[diesel(belongs_to(Db_Card, foreign_key = card_id))]
#[diesel(check_for_backend(diesel::sqlite::Sqlite))]
#[diesel(primary_key(deck_id, card_id))]
pub struct Db_DeckCard {
  #[builder(start_fn)]
  pub deck_id: Db_DeckId,

  #[builder(start_fn)]
  pub card_id: Db_CardLocalId,

  #[builder(default, into)]
  pub main: Db_DeckCardAmount,

  #[builder(default, into)]
  pub extra: Db_DeckCardAmount,

  #[builder(default, into)]
  pub side: Db_DeckCardAmount,
}

impl Db_DeckCard {
  #[inline]
  pub fn id(&self) -> Db_DeckCardId {
    Db_DeckCardId {
      deck_id: self.deck_id,
      card_id: self.card_id,
    }
  }
}

#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash, Serialize, Deserialize, Type)]
pub struct Db_DeckCardId {
  pub deck_id: Db_DeckId,
  pub card_id: Db_CardLocalId,
}

File: app/src/lib/model/deck-card.ts
import type { Db_CardLocalId, Db_DeckCard, Db_DeckCardAmount, Db_DeckId } from "@/lib/bindings";

export class DeckCardImpl implements Db_DeckCard {
  public readonly deck_id: Db_DeckId;
  public readonly card_id: Db_CardLocalId;
  public main: Db_DeckCardAmount;
  public extra: Db_DeckCardAmount;
  public side: Db_DeckCardAmount;

  constructor(card: Db_DeckCard) {
    this.deck_id = card.deck_id;
    this.card_id = card.card_id;
    this.main = card.main;
    this.extra = card.extra;
    this.side = card.side;
  }

  public isEmpty() {
    return this.main < 1 && this.extra < 1 && this.side < 1;
  }

  public toJSON(): Db_DeckCard {
    return {
      deck_id: this.deck_id,
      card_id: this.card_id,
      main: this.main,
      extra: this.extra,
      side: this.side,
    };
  }
}
